scripts/data_prep/convert_text_to_mds.py

```diff
@@ -0,0 +1,352 @@
+    if object_store is not None:
+        _, _, folder_prefix = parse_uri(input_folder)
+        names = [
+            name for name in object_store.list_objects(folder_prefix)
+            if name.endswith('.txt')
+        ]
+    else:
+        # input_folder is a local folder
+        names = [
+            text_file for dirpath, _, _ in os.walk(input_folder)
+            for text_file in glob(os.path.join(dirpath, '*.txt'))
+        ]
+    # return names, sizes
+    print(f'Found {len(names)} text files at {input_folder}')
+
+    return names
+
+
+def get_task_args(
+    object_names: List[str],
+    output_root: str,
+    input_folder: str,
+    n_groups: int,
+    tokenizer_name: str,
+    concat_tokens: int,
+    eos_text: str,
+    bos_text: str,
+    no_wrap: bool,
+    compression: str,
+    max_mds_writer_workers: int,
+) -> Iterable:
+    num_objects = len(object_names)
+    objs_per_group = math.ceil(num_objects / n_groups)
+    for group, i in enumerate(range(0, num_objects, objs_per_group)):
+        output_subdir = os.path.join(output_root, str(group))
+        yield (
+            object_names[i:min(i + objs_per_group, num_objects)],
+            output_subdir,
+            input_folder,
+            tokenizer_name,
+            concat_tokens,
+            eos_text,
+            bos_text,
+            no_wrap,
+            compression,
+            max_mds_writer_workers,
+        )
+
+
+def download_and_convert_starargs(args: Tuple):
+    return download_and_convert(*args)
+
+
+def download_and_convert(
+    file_names: List[str],
+    output_folder: str,
+    input_folder: str,
+    tokenizer_name: str,
+    concat_tokens: int,
+    eos_text: str,
+    bos_text: str,
+    no_wrap: bool,
+    compression: str,
+    max_mds_writer_workers: int,
+):
+    object_store = maybe_create_object_store_from_uri(input_folder)
+
+    # Download file_names
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        downloading_iter = DownloadingIterable(object_names=file_names,
+                                               output_folder=tmp_dir,
+                                               object_store=object_store)
+        tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
+        tokenizer.model_max_length = 5000000000  # Hack to prevent warnings from HuggingFace
+
+        # Use the ConcatTokensDataset from LLM-foundry to concatenate sequences of tokens up to the maximum sequence length
+        dataset = ConcatTokensDataset(
+            hf_dataset=downloading_iter,
+            max_length=concat_tokens,
+            tokenizer=tokenizer,
+            eos_text=eos_text,
+            bos_text=bos_text,
+            no_wrap=no_wrap,
+        )
+
+        columns = {'tokens': 'bytes'}
+
+        print(f'Converting to MDS format...')
+        with MDSWriter(out=output_folder,
+                       columns=columns,
+                       max_mds_writer_workers=max_mds_writer_workers,
```

Comment:
[MDSWriter()](https://docs.mosaicml.com/projects/streaming/en/stable/api_reference/generated/streaming.MDSWriter.html#streaming.MDSWriter) does not have a `max_mds_writer_workers` parameter !!!.

You should instantiate one `MDSWriter` per process and ensure each process has a unique `output_folder`. Happy to explain you offline on how it works.

Reply:
@karan6181 seems like this should have crashed. Why does mdswriter accept arbitrary kwargs?

Reply:
ah yes, good catch! that was from a refactor. MDSWriter takes arbitrary kwargs, so it wasn't caught at runtime. My intention was for that to be max_workers=max_mds_writer_workers. Is this correct? It seems like that might not be necessary because MDSWriter aways writes to a local output folder here and max_workers is stated in the docstring to be for remote uploads. 

> You should instantiate one MDSWriter per process and ensure each process has a unique output_folder. Happy to explain you offline on how it works.


Currently there is one MDSWriter and a unique output folder per process. In `get_task_args` we assign unique subdirectory per process.

Reply:
> seems like this should have crashed. Why does mdswriter accept arbitrary kwargs?

MDSWriter uses kwargs `progress_bar` and `max_workers`. Looks like streaming does not do a strict validation on the incorrect kwargs. I will take a look at it. 

>Currently there is one MDSWriter and a unique output folder per process. In get_task_args we assign unique subdirectory per process.

Make sense. Thank You!

Reply:
Removed the max_workers argument because it is only used for remote uploads and we are writing to a local dir.